### DISCUSSION TITLE
Is the debugging inside `PickSendVote` thread safe?

### DISCUSSION THREAD

**Discussion by tnasu** [tendermint/tendermint#9353] [Q&A]
I'm not sure that but, is this debugging of `PeerState` thread safe?
* https://github.com/tendermint/tendermint/blob/main/consensus/reactor.go#L1075
<img width="735" alt="screenshot-2022-09-02 10 54 56" src="https://user-images.githubusercontent.com/1164270/188042218-0c862eb5-9335-45b6-8717-e4843b6d33c7.png">

Of course, I know the debug level logging doesn't usually run with nodes. But, sometimes, I want to run with debug level and I met the panic on this logging.

Thanks

**Comment by torao** [comment]
Not an answer, but a supplement to this question, `PeerState` always does `ps.mtx.Lock()` when updating and referencing the state, but in this `logger.Debug()` an unlocked `PeerState` instance is passed. The underlying zerolog Logger in Cosmos SDK tries to reference all fields of the `PeerState` (whether locked or not) in the reflection to marshal to JSON.

Since pointer address updates are not an atomic operation, if the debug logging is enabled in a multi-node with high-load environment, would this `ps.logger.Debug()` contains a potential problem with reading out invalid pointer addresses during updates (probably causing SIGSEGV)?
